File: src/sequence_notes.py
import re
import csv
import unicodedata
from collections import defaultdict
import sys
import tempfile
import os
import shutil

# ============================================================
# HEBREW NORMALIZATION
# ============================================================

def normalize_hebrew(text):
    """
    Remove Hebrew vowels, accents, joiners, and normalize.
    """
    text = unicodedata.normalize("NFD", text)

    chars = []
    for ch in text:
        name = unicodedata.name(ch, "")

        if (
            "HEBREW POINT" in name
            or "HEBREW ACCENT" in name
            or "MASORA" in name
        ):
            continue

        if ch in ("\u2060", "\u200d", "\ufeff"):
            continue

        chars.append(ch)

    text = "".join(chars)

    text = text.replace("־", "")
    text = text.replace("⁠", "")

    return unicodedata.normalize("NFC", text).strip()


# ============================================================
# PARSE USFM ALIGNMENTS
# ============================================================

def build_alignment_map(usfm_path):
    """
    Returns:

    verse_map["10:1"] = [
        ("ישראל", 1),
        ("בוקק", 2),
        ("גפן", 4),
        ...
    ]
    """

    with open(usfm_path, encoding="utf-8") as f:
        usfm = f.read()

    verse_map = defaultdict(list)

    chapter = None
    verse = None

    token_pattern = re.compile(
        r'\\c\s+(\d+)'
        r'|\\v\s+(\d+)'
        r'|\\zaln-s\b.*?x-content="([^"]+)"'
        r'|\\w\s+([^|\\]+?)\|',
        re.DOTALL,
    )

    english_position = 0
    pending_hebrew = []

    for match in token_pattern.finditer(usfm):

        chap, vrs, hebrew, english = match.groups()

        if chap:
            chapter = chap
            continue

        if vrs:
            verse = f"{chapter}:{vrs}"
            english_position = 0
            pending_hebrew = []
            continue

        if hebrew:
            pending_hebrew.append(normalize_hebrew(hebrew))
            continue

        if english:
            english_position += 1

            if pending_hebrew:
                for hw in pending_hebrew:
                    verse_map[verse].append((hw, english_position))

                pending_hebrew = []

    return verse_map


# ============================================================
# FIND QUOTE POSITION
# ============================================================

def quote_position(quote, verse_alignments):
    """
    verse_alignments:
        [
            ("ישראל", 1),
            ("בוקק", 2),
            ...
        ]

    Returns:
        (position, length)
    """

    quote_parts = [p.strip() for p in quote.split("&")]

    positions = []
    total_words = 0

    hebrew_sequence = [h for h, _ in verse_alignments]
    english_positions = [p for _, p in verse_alignments]

    for part in quote_parts:

        words = [
            normalize_hebrew(w)
            for w in part.split()
            if normalize_hebrew(w)
        ]

        total_words += len(words)

        if not words:
            continue

        found = False

        for i in range(len(hebrew_sequence) - len(words) + 1):

            if hebrew_sequence[i:i + len(words)] == words:

                positions.append(english_positions[i])
                found = True
                break

        if not found:
            # fall back to first word only
            first_word = words[0]

            for i, hw in enumerate(hebrew_sequence):
                if hw == first_word:
                    positions.append(english_positions[i])
                    found = True
                    break

    if not positions:
        return None, total_words

    return min(positions), total_words


# ============================================================
# SORT NOTES
# ============================================================

def sort_notes(tsv_file, verse_map):

    with open(tsv_file, encoding="utf-8") as f:
        reader = csv.reader(f, delimiter="\t")
        header = next(reader)
        rows = list(reader)

    grouped = defaultdict(list)

    for original_index, row in enumerate(rows):

        if not row:
            continue

        reference = row[0]

        grouped[reference].append(
            {
                "row": row,
                "index": original_index,
            }
        )

    sorted_rows = []

    for reference in grouped:

        verse_notes = grouped[reference]

        decorated = []

        for note_index, note in enumerate(verse_notes):

            row = note["row"]

            note_id = row[1]

            quote = row[4].strip() if len(row) > 4 else ""

            verse_alignments = verse_map.get(reference, [])

            pos, quote_len = quote_position(
                quote,
                verse_alignments,
            )

            if pos is None:

                print(f"UNMATCHED: {reference}  {note_id}")

                sort_key = (
                    float("inf"),
                    0,
                    note_index,
                )

            else:

                sort_key = (
                    pos,
                    -quote_len,
                    note_index,
                )

            decorated.append(
                (
                    sort_key,
                    row,
                )
            )

        decorated.sort(key=lambda x: x[0])

        sorted_rows.extend(row for _, row in decorated)

    def safe_write_tsv(tsv_file, rows):
        # 1. Write to temp file in same directory
        dir_name = os.path.dirname(tsv_file)

        with tempfile.NamedTemporaryFile(
            mode="w",
            encoding="utf-8",
            delete=False,
            dir=dir_name,
            newline="",
        ) as tmp:
            writer = csv.writer(
                tmp,
                delimiter="\t",
                lineterminator="\n",
                quoting=csv.QUOTE_MINIMAL,
            )
            writer.writerows(rows)
            temp_path = tmp.name

        # 2. Backup old file (only AFTER successful write)
        if os.path.exists(tsv_file):
            backup_file = tsv_file + ".old"
            shutil.move(tsv_file, backup_file)

        # 3. Atomic replace
        shutil.move(temp_path, tsv_file)

    safe_write_tsv(tsv_file, [header] + sorted_rows)

# ============================================================
# MAIN
# ============================================================

def sequence_notes(ult_usfm, notes_tsv):

    print("Reading alignments...")
    verse_map = build_alignment_map(ult_usfm)

    print("Sorting notes...")
    sort_notes(
        notes_tsv,
        verse_map,
    )

    print(f"Done. Output written to {notes_tsv}")


if __name__ == "__main__":
    if len(sys.argv) != 3:
        raise ValueError("Usage: script.py ult_usfm notes_tsv")

    ult_usfm = sys.argv[1]
    notes_tsv = sys.argv[2]

    sequence_notes(ult_usfm, notes_tsv)

File: src/lib/sequence-notes.js
const fs = require('fs');
const path = require('path');

function normalizeHebrew(text) {
  if (!text) return '';
  return text
    .normalize('NFD')
    .replace(/[\u0591-\u05BD\u05BF\u05C1-\u05C2\u05C4-\u05C5\u05C7]/g, '')
    .replace(/[\u2060\u200d\ufeff]/g, '')
    .replace(/[\u05be\u2060]/g, '')
    .normalize('NFC')
    .trim();
}

function buildAlignmentMap(usfmPath) {
  const usfm = fs.readFileSync(usfmPath, 'utf8');
  const verseMap = {};

  let chapter = null;
  let verse = null;
  let englishPosition = 0;
  let pendingHebrew = [];

  const tokenPattern = /\\c\s+(\d+)|\\v\s+(\d+)|\\zaln-s\b.*?x-content="([^"]+)"|\\w\s+([^|\\]+?)\|/gs;
  let match;
  while ((match = tokenPattern.exec(usfm)) !== null) {
    const [, chap, vrs, hebrew, english] = match;

    if (chap) {
      chapter = chap;
      continue;
    }

    if (vrs) {
      verse = `${chapter}:${vrs}`;
      if (!verseMap[verse]) verseMap[verse] = [];
      englishPosition = 0;
      pendingHebrew = [];
      continue;
    }

    if (hebrew) {
      pendingHebrew.push(normalizeHebrew(hebrew));
      continue;
    }

    if (english) {
      englishPosition += 1;
      if (verse && pendingHebrew.length) {
        for (const hw of pendingHebrew) {
          verseMap[verse].push([hw, englishPosition]);
        }
        pendingHebrew = [];
      }
    }
  }

  return verseMap;
}

function quotePosition(quote, verseAlignments) {
  const quoteParts = String(quote || '').split('&').map((part) => part.trim());
  const hebrewSequence = verseAlignments.map(([hebrew]) => hebrew);
  const englishPositions = verseAlignments.map(([, position]) => position);

  const positions = [];
  let totalWords = 0;

  for (const part of quoteParts) {
    const words = part
      .split(/\s+/)
      .map((word) => normalizeHebrew(word))
      .filter(Boolean);

    totalWords += words.length;
    if (!words.length) continue;

    let found = false;
    for (let i = 0; i <= hebrewSequence.length - words.length; i++) {
      let matches = true;
      for (let j = 0; j < words.length; j++) {
        if (hebrewSequence[i + j] !== words[j]) {
          matches = false;
          break;
        }
      }
      if (matches) {
        positions.push(englishPositions[i]);
        found = true;
        break;
      }
    }

    if (!found) {
      const firstWord = words[0];
      for (let i = 0; i < hebrewSequence.length; i++) {
        if (hebrewSequence[i] === firstWord) {
          positions.push(englishPositions[i]);
          break;
        }
      }
    }
  }

  if (!positions.length) return [null, totalWords];
  return [Math.min(...positions), totalWords];
}

function getQuote(row) {
  const cols = Array.isArray(row) ? row : String(row || '').split('\t');
  return (cols[4] || '').trim();
}

function getSequenceSortKey(row, verseMap) {
  const cols = Array.isArray(row) ? row : String(row || '').split('\t');
  const reference = cols[0] || '';
  const verseAlignments = verseMap[reference] || [];
  const [position, quoteLength] = quotePosition(getQuote(cols), verseAlignments);
  if (position == null) return [Infinity, 0];
  return [position, -quoteLength];
}

function sortRowsBySequence(rows, verseMap) {
  const grouped = new Map();
  rows.forEach((row, index) => {
    const cols = Array.isArray(row) ? row : String(row || '').split('\t');
    const reference = cols[0] || '';
    if (!grouped.has(reference)) grouped.set(reference, []);
    grouped.get(reference).push({ row, index });
  });

  const sortedRows = [];
  for (const verseNotes of grouped.values()) {
    const decorated = verseNotes.map((note, noteIndex) => ({
      row: note.row,
      sortKey: [...getSequenceSortKey(note.row, verseMap), noteIndex],
    }));

    decorated.sort((a, b) => {
      for (let i = 0; i < a.sortKey.length; i++) {
        if (a.sortKey[i] !== b.sortKey[i]) return a.sortKey[i] - b.sortKey[i];
      }
      return 0;
    });

    sortedRows.push(...decorated.map(({ row }) => row));
  }

  return sortedRows;
}

function parseTsvLine(line) {
  const cells = [];
  let cell = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      if (inQuotes && line[i + 1] === '"') {
        cell += '"';
        i += 1;
      } else {
        inQuotes = !inQuotes;
      }
    } else if (ch === '\t' && !inQuotes) {
      cells.push(cell);
      cell = '';
    } else {
      cell += ch;
    }
  }

  cells.push(cell);
  return cells;
}

function escapeTsvCell(cell) {
  const value = cell == null ? '' : String(cell);
  if (!/["\t\r\n]/.test(value)) return value;
  return `"${value.replace(/"/g, '""')}"`;
}

function writeTsvAtomic(tsvFile, rows) {
  const dirName = path.dirname(tsvFile);
  const tempPath = path.join(dirName, `.sequence-notes-${process.pid}-${Date.now()}-${Math.random().toString(36).slice(2)}.tmp`);
  const content = rows.map((row) => row.map(escapeTsvCell).join('\t')).join('\n') + '\n';
  fs.writeFileSync(tempPath, content, 'utf8');
  if (fs.existsSync(tsvFile)) {
    fs.renameSync(tsvFile, `${tsvFile}.old`);
  }
  fs.renameSync(tempPath, tsvFile);
}

function sequenceNotes(ultUsfm, notesTsv) {
  const verseMap = buildAlignmentMap(ultUsfm);
  const content = fs.readFileSync(notesTsv, 'utf8');
  const lines = content.split(/\r?\n/).filter((line) => line.length);
  if (!lines.length) throw new Error('Notes TSV is empty');

  const header = parseTsvLine(lines[0]);
  const rows = lines.slice(1).map(parseTsvLine);
  const sortedRows = sortRowsBySequence(rows, verseMap);
  writeTsvAtomic(notesTsv, [header, ...sortedRows]);
}

module.exports = {
  buildAlignmentMap,
  getSequenceSortKey,
  normalizeHebrew,
  quotePosition,
  sequenceNotes,
  sortRowsBySequence,
};
